src/node/src/indexer_impl.rs

```diff
@@ -99,77 +97,132 @@ impl IndexerImpl {
-                                            _ => {
-                                                let msg = format!("lock pending_databases failed");
-                                                return Err(Status::internal(msg));
-                                            }
-                                        },
-                                        Err(e) => {
-                                            let msg = format!("{e}");
-                                            return Err(Status::internal(msg));
-                                        }
-                                    }
-                                }
-                                PayloadType::QuerySessionPayload => {
-                                    info!("Skip QuerySessionPayload");
-                                }
-                                PayloadType::MintCreditsPayload => {
-                                    info!("Skip MintCreditsPayload");
+                debug!("Block transaction size: {}", block.data.len());
+                match self.node_store.lock() {
+                    Ok(mut store) => {
+                        store.get_auth_store().begin_block(
+                            block.header.height.value(),
+                            Utc::now().timestamp() as u64,
+                        );
+                        let mut pending_databases: Vec<(AccountAddress, DatabaseMutation, TxId)> =
+                            vec![];
+                        Self::parse_and_pending_mutations(&block.data, &mut pending_databases)
+                            .await?;
+                        Self::apply_database_mutations(
+                            store.get_auth_store(),
+                            &mut pending_databases,
+                        )
+                        .await?;
+                        store.get_auth_store().commit().map_err(|e| {
+                            Status::internal(format!("fail to commit database for {e}"))
+                        })?;
+                        // TODO: add show indexer state api
+                        info!(
+                            "[Indexer] last block state: {:?}",
+                            store.get_auth_store().get_last_block_state()
+                        );
+                        info!(
+                            "[Indexer] indexer state: {:?}",
+                            store.get_auth_store().get_state()
+                        );
+                    }
+                    Err(e) => {
+                        warn!("[Indexer] get node store error: {:?}", e);
+                        return Err(Status::internal(format!("get node store error: {:?}", e)));
+                    }
+                }
+            }
+            _ => {}
+        }
+        Ok(())
+    }
+    async fn parse_and_pending_mutations(
+        txs: &Vec<Vec<u8>>,
+        database_mutations: &mut Vec<(AccountAddress, DatabaseMutation, TxId)>,
```

Comment:
why not return database mutations directly?

Reply:
account_addr and tx_id are required for apply_database. This logic is borrowed from abci_impl. I consider to extract common function for abci and indexer_impl.

Reply:
https://github.com/dbpunk-labs/db3/blob/9eec75e62585ce39a1efbc45ad15e0696e4f3f13/src/node/src/abci_impl.rs#L392-L393